internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -532,16 +532,20 @@ func (r *Reconciler) updateMachineDeployment(ctx context.Context, d *clusterv1.M
 
 // We have this as standalone variant to be able to use it from the tests.
 func updateMachineDeployment(ctx context.Context, c client.Client, d *clusterv1.MachineDeployment, modify func(*clusterv1.MachineDeployment)) error {
+	mdObjectKey := util.ObjectKey(d)
 	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
-		if err := c.Get(ctx, util.ObjectKey(d), d); err != nil {
+		// Note: We intentionally don't re-use the passed in MachineDeployment d here as that would
+		// overwrite any local changes we might have previously made to the MachineDeployment with the version
+		// we get here from the apiserver.
+		md := &clusterv1.MachineDeployment{}
+		if err := c.Get(ctx, mdObjectKey, md); err != nil {
```

Comment:
Out of curiosity: when we update the MachineDeployment here and dont store the new version in `d`, at the end of the reconcile loop when we run `patchMachineDeployment` it would run against a MachineDeployment that has an older resourceVersion than the one on the API server, correct? That should cause a problem, right?

@sbueringer I remember that you talked about a fix in the patch helper that could involve the resourceVersion and including safeguards so that the patch does not run against an old version of the object. I see that the pipelines on the PR are green. Is this because the patch helper currently does not have the "resource version fix" or am misunderstanding the flow here?

Reply:
Yes. Overall Patchhelper doesn't fail when the resourceversion doesn't match. For some patch calls it doesn't include it, for others it retries with the latest version from the API.

Reply:
>// Note: We intentionally don't re-use the passed in MachineDeployment d here as that would
		// overwrite any local changes we might have previously made to the MachineDeployment with the version
		// we get here from the apiserver

I'm not sure I'm following the comment. The cr version is dicated by the client not the server.

Reply:
Do you mean CR apiVersion or resourceVersion with "cr version"?

What I tried to say with that comment is:

If we use the passed in `d` paramter and then run this `Get` call whatever content `d` previously had is then overwritten with what we Get from the client (most precisely from the cache). And this change is propagated back up the call stack.

For example:

* Reconcile func
  * Making some changes to d.spec...refs
  * calling updateMachineDeployment
    * Get will replace d with what we "Get' from the Client, changes to the refs from above are lost
    * Because we're using pointers the changes to d will also affect the func calling updateMachineDeployment


Happy to rephrase the comment to make this clearer